test/e2e/reconciler/namespace/controller_test.go

```diff
@@ -126,6 +136,100 @@ func TestNamespaceScheduler(t *testing.T) {
+				require.NoError(t, err, "failed to create cluster")
+				require.Eventually(t, func() bool {
+					cluster, err = server.kcpClient.WorkloadV1alpha1().WorkloadClusters().Get(ctx, cluster.Name, metav1.GetOptions{})
+					conditions.MarkTrue(cluster, "Ready")
+					cluster, err = server.kcpClient.WorkloadV1alpha1().WorkloadClusters().UpdateStatus(ctx, cluster, metav1.UpdateOptions{})
+					return err == nil
+				}, wait.ForeverTestTimeout, time.Millisecond*100)
+				go wait.UntilWithContext(ctx, func(ctx context.Context) {
+					patchBytes := []byte(fmt.Sprintf(`[{"op":"replace","path":"/status/lastSyncerHeartbeatTime","value":%q}]`, time.Now().Format(time.RFC3339)))
+					_, err := server.kcpClient.WorkloadV1alpha1().WorkloadClusters().Patch(ctx, cluster.Name, types.JSONPatchType, patchBytes, metav1.PatchOptions{}, "status")
+					if err != nil {
+						klog.Errorf("failed to set status.lastSyncerHeartbeatTime: %v", err)
+						return
+					}
+				}, time.Second*10)
+
+				t.Log("Create a new unique sheriff CRD")
+				group := uuid.New().String() + ".io"
+				crd := newSheriffCRD(group)
+				gvr := schema.GroupVersionResource{Group: crd.Spec.Group, Resource: "sheriffs", Version: "v1"}
+				err = configcrds.CreateSingle(ctx, crdClusterClient.Cluster(server.clusterName).ApiextensionsV1().CustomResourceDefinitions(), crd)
+				require.NoError(t, err, "error bootstrapping CRD %s in cluster %s", crd.Name, server.clusterName)
+				require.Eventually(t, func() bool {
+					_, err := dynamicClusterClient.Cluster(server.clusterName).Resource(gvr).Namespace("").List(ctx, metav1.ListOptions{})
+					return err == nil
+				}, wait.ForeverTestTimeout, time.Millisecond*100, "failed to see CRD in cluster")
+
+				t.Log("Create a sheriff and wait for it to be scheduled")
+				_, err = dynamicClusterClient.Cluster(server.clusterName).Resource(gvr).Namespace("default").Create(ctx, newSheriff(group, "woody"), metav1.CreateOptions{})
+				require.NoError(t, err, "failed to create sheriff")
+				require.Eventually(t, func() bool {
+					obj, err := dynamicClusterClient.Cluster(server.clusterName).Resource(gvr).Namespace("default").Get(ctx, "woody", metav1.GetOptions{})
+					if err != nil {
+						return false
+					}
+					require.NoError(t, err, "failed to get sheriff")
+					return obj.GetLabels()[clusterLabel] != ""
+				}, wait.ForeverTestTimeout, time.Millisecond*100, "failed to see sheriff scheduled")
+
+				t.Log("Delete the sheriff and the sheriff CRD")
+				err = dynamicClusterClient.Cluster(server.clusterName).Resource(gvr).Namespace("default").Delete(ctx, "woody", metav1.DeleteOptions{})
+				require.NoError(t, err, "failed to delete sheriff")
+				err = crdClusterClient.Cluster(server.clusterName).ApiextensionsV1().CustomResourceDefinitions().Delete(ctx, crd.Name, metav1.DeleteOptions{})
+				require.NoError(t, err, "failed to delete CRD")
+
+				time.Sleep(30 * time.Second)
```

Comment:
Ouch. Could we maybe instead poll for the API group to be gone from discovery, and then maybe add a delay (if needed) to hopefully be certain the namespace scheduler has stopped the dynamic informer? (Or maybe we need some way to read this information?)

Reply:
yep, is ugly. But these are to be sure the dynamic informer has stopped for the gvr. We could probably reduce it to 10s or so because we have a 5s poll interval. Any other idea?

Reply:
Poll discovery for it to be gone, then sleep 10s?

Reply:
reduced to 7s > 5s.